### ISSUE TITLE
[Question] Replace OpenCv default dlls with other OpenCv build

### ISSUE THREAD

**Issue by JuanDYB** [emgucv/emgucv#924]
OpenCv for windows is not build with GStremer backend. 
I have follow the steps to build OpenCv for Windows with Gstreamer backend because I want to do a RTSP streaming server to publish `Mat` frames as video over ethernet.
How can I replace OpenCv Dlls that comes with EmguCV with the OpenCv dlls that I have build with Gstreamer.

With regards,

**Comment by sn4k3** [comment]
You just need to replace all dlls for output emgucv compilation. I suspect you only need to replace the cvextern.dll

**Comment by sn4k3** [comment]
You can use openCV compilation too, copy the outputs to your project bin root. And this way not not use the emgucv runtime dist.

**Comment by JuanDYB** [comment]
I don't understand so well. You mean install EmguCv nugget and after that copy binaries from OpenCv Build?
I have try this and didn't work. It sais that cvextern.dll is not found.

**Comment by sn4k3** [comment]
Yes, just install the emgucv nugget and copy the dlls. It should work on bin root. However in a normal not published compilation they are located under `runtimes\win-x64\native`

```
 Directory: UVtools\UVtools.UI\bin\Debug\net6.0\runtimes\win-x64\native

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          19/09/2020    01:46        4369408 av_libglesv2.dll
-a---          18/10/2023    01:41         327576 concrt140.dll
-a---          16/09/2023    14:03       35899904 cvextern.dll
-a---          20/09/2023    19:11        1607088 libHarfBuzzSharp.dll
-a---          10/01/2024    22:47        9414064 libSkiaSharp.dll
-a---          18/10/2023    01:41          35704 msvcp140_1.dll
-a---          18/10/2023    01:41         267160 msvcp140_2.dll
-a---          18/10/2023    01:41          50072 msvcp140_atomic_wait.dll
-a---          18/10/2023    01:41          31640 msvcp140_codecvt_ids.dll
-a---          18/10/2023    01:41         578384 msvcp140.dll
-a---          17/10/2023    23:37       26353152 opencv_videoio_ffmpeg481_64.dll
-a---          18/10/2023    01:41          49560 vcruntime140_1.dll
-a---          18/10/2023    01:41         109440 vcruntime140.dll
```

**Comment by JuanDYB** [comment]
Hi,
It seems that I don't have the same dlls.
If I install `EmguCv` nugget I get this dlls in `bin/Debug` directory.

**bin/Debug**
```
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----        06/04/2024     22:28                runtimes
-a----        18/10/2023      3:37          24520 Emgu.CV.Bitmap.dll
-a----        18/10/2023      2:59         891336 Emgu.CV.dll
```
And after that I have created runtimes directory as you mentioned and I have put all the dlls that I have in `install/x64/vc17/bin` folder of OpenCv output.
By doing this I'm getting `DllNotFoundException` because it seems EmguCv is searching for a dll named `cvextern.dll`
**bin/Debug/runtime/win-x64/native**
```
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----        04/04/2024     16:32        3082752 opencv_calib3d490.dll
-a----        04/04/2024     16:11        7384064 opencv_calib3d490d.dll
-a----        04/04/2024     16:28       14705664 opencv_core490.dll
-a----        04/04/2024     16:10       23786496 opencv_core490d.dll
-a----        04/04/2024     16:34        7255040 opencv_dnn490.dll
-a----        04/04/2024     16:12       17175040 opencv_dnn490d.dll
-a----        04/04/2024     16:30         905216 opencv_features2d490.dll
-a----        04/04/2024     16:11        2667520 opencv_features2d490d.dll
-a----        04/04/2024     16:29         585728 opencv_flann490.dll
-a----        04/04/2024     16:10        1609216 opencv_flann490d.dll
-a----        04/04/2024     16:40        6283264 opencv_gapi490.dll
-a----        04/04/2024     16:14       17370624 opencv_gapi490d.dll
-a----        04/04/2024     16:32         346112 opencv_highgui490.dll
-a----        04/04/2024     16:12         910848 opencv_highgui490d.dll
-a----        04/04/2024     16:30        3586048 opencv_imgcodecs490.dll
-a----        04/04/2024     16:11        8611840 opencv_imgcodecs490d.dll
-a----        04/04/2024     16:30       29279232 opencv_imgproc490.dll
-a----        04/04/2024     16:11       45126656 opencv_imgproc490d.dll
-a----        04/04/2024     16:29         670208 opencv_ml490.dll
-a----        04/04/2024     16:10        1698304 opencv_ml490d.dll
-a----        04/04/2024     16:38        1435136 opencv_objdetect490.dll
-a----        04/04/2024     16:13        3610112 opencv_objdetect490d.dll
-a----        04/04/2024     16:30         830464 opencv_photo490.dll
-a----        04/04/2024     16:11        1881600 opencv_photo490d.dll
-a----        04/04/2024     16:32        1001984 opencv_stitching490.dll
-a----        04/04/2024     16:12        2457600 opencv_stitching490d.dll
-a----        04/04/2024     16:37         836608 opencv_video490.dll
-a----        04/04/2024     16:13        2070528 opencv_video490d.dll
-a----        04/04/2024     16:32         844288 opencv_videoio490.dll
-a----        04/04/2024     16:11        2054656 opencv_videoio490d.dll
-a----        02/04/2024      7:55       26314752 opencv_videoio_ffmpeg490_64.dll
```

**Comment by sn4k3** [comment]
Theres not an option to combine all that into a single dll?
If you want emgucv runtime you need to get the package https://www.nuget.org/packages/Emgu.CV.runtime.windows

**Comment by JuanDYB** [comment]
I didn't see any cmake option to build as single dll.
I wan't to use my OpenCv build because it seems that default compilation of OpenCv doesn't include Gstreamer backend. That's why I'm trying to do in that way.

**Comment by sn4k3** [comment]
Better ask to @emgucv 
But you can also edit the [bat](https://github.com/emgucv/emgucv/blob/master/platforms/windows/Build_Binary_x86.bat) and add `WITH_GSTREAMER:BOOL=TRUE`

**Comment by JuanDYB** [comment]
I think it's more complicated, I'll try as you say but building OpenCv with GStreamer require to fill GStremer paths and other variables. Using CMake for windows and building winth VisualStudio it seems that works (show GStreamer in GetBuildInformation() method).

Thanks for your help.

**Comment by sn4k3** [comment]
See: https://discourse.cmake.org/t/how-to-combine-all-dlls-from-subproject-into-one-dll/7235/5
`BUILD_opencv_world` which emgucv is using [here](https://github.com/emgucv/emgucv/blob/master/platforms/windows/Build_Binary_x86.bat#L523)

**Comment by JuanDYB** [comment]
You are sugesting to use this cmake option to build OpenCv and rename as cvextern.dll?

**Comment by sn4k3** [comment]
Yes, it should combine all in one dll and then you can try it as cvextern.dll

**Comment by JuanDYB** [comment]
Hi,

I have tried to rebuild Emgucv from scratch following documentation.

1. Clone git repository
2. Executing `Build_Binary_x86-64_doc.bat`
3. Open build_x86_64 folder with cmake, setting new variable `GSTREAMER_DIR=C:/gstreamer/1.0/msvc_x86_64`
4. Click on configure and cmake fill `WITH_GSTREAMER=ON` and all the Gstreamer options paths.
5. Click on generate
6. Open `build_x86_64/emgucv.sln`. Set Release mode and rebuild cvextern project.
7. Copy libs/runtimes directory to new VisualStudio project bin/Degug (project has EmguCv nugget installed)

It seems it isn't working. At starting I have called `CvInvoke.BuildInformation` and it is throwing `System.TypeLoadException`

**Comment by emgucv** [comment]
You will also need to copy all the GStreamer dlls to the folder that contains cvextern.dll. 

If you still got the System.TypeLoadException after that, follow this instruction to use dependecy walker to find out what dependecy is missing from the folder of the dlls:
https://www.emgu.com/wiki/index.php/Download_And_Installation#Are_you_missing_any_dependency.3F

If dependency walker reports missing dependency, copy them over to the folder where you have the "cvextern.dll".

**Comment by JuanDYB** [comment]
Finally I did it!
I copied all dlls in `C:/gstreamer/1.0/msvc_x86_64/bin` to `runtimes` dir inside `bin`, at same level as `cvextern.dll`.
Now it seems to work, and I see `GSTREAMER` enabled in `getBuildInformation()`.

But I'm having problems to build Gstreamer pipeline. I don't know if maybe `gstreamer` needs more dependencies or somethig like that.

`string command = "appsrc ! videoconvert ! mfxh264enc ! video/x-h264, stream-format=byte-stream, profile=baseline ! send-config=true rtph264pay pt=96 ! udpsink host=127.0.0.1 port=5004";`

```
[ WARN:0@0.957] global cap_gstreamer.cpp:2250 cv::CvVideoWriter_GStreamer::close_ OpenCV | GStreamer warning: No source in GStreamer pipeline. Ignore
[ERROR:0@0.960] global cap.cpp:645 cv::VideoWriter::open VIDEOIO(GSTREAMER): raised OpenCV exception:
 
OpenCV(4.9.0) C:\CMake\opencv\opencv\modules\videoio\src\cap_gstreamer.cpp:154: error: (-215:Assertion failed) ptr in function 'cv::`anonymous-namespace'::GSafePtr<struct _GstElement>::get'
```

Thanks for your support.

**Comment by emgucv** [comment]
Closing this as resolved — the main question was answered and confirmed working in the thread. Summarizing the working recipe for anyone needing a custom-configured OpenCV (e.g. GStreamer on Windows) with Emgu CV, since the standard packages statically link OpenCV into `cvextern.dll` and its parts cannot be swapped individually:

1. Clone the repo matching your Emgu CV release and run `platforms\windows\Build_Binary_x86-64_doc.bat` once to set up the CMake tree.
2. In the `build_x86_64` CMake configuration, set your custom options (here: `GSTREAMER_DIR=C:/gstreamer/1.0/msvc_x86_64`, which enables `WITH_GSTREAMER`), then configure + generate.
3. Build the `cvextern` project from `build_x86_64\emgucv.sln` in Release.
4. Replace the `cvextern.dll` in your application with the newly built one, and copy the extra native dependencies (for GStreamer: everything in `C:\gstreamer\1.0\msvc_x86_64\bin`) next to it. `CvInvoke.BuildInformation` confirms the enabled backends; Dependency Walker helps find any still-missing DLL.

The remaining discussion about assembling the GStreamer pipeline string itself is GStreamer usage rather than Emgu CV — the `gst-launch-1.0` tool is the best way to prototype a pipeline before using it in `VideoWriter`. Closing as answered.
